File: iOS/YouKong/Presentation/Screens/Invitation/CreateInvitationViewModel.swift
//
//  CreateInvitationViewModel.swift
//  YouKong
//
//  创建邀请 ViewModel
//

import Foundation
import Combine
import Factory

@MainActor
final class CreateInvitationViewModel: ObservableObject {
    @Published var circles: [FriendCircle] = []
    @Published var selectedCircleId: String?
    @Published var maxUses: Int = 100
    @Published var expiresDays: Int = 7
    @Published var isLoading = false
    @Published var isCreating = false
    @Published var errorMessage: String?
    @Published var createdInvitation: InvitationDetail?

    @Injected(\.circleRepository) private var circleRepository
    @Injected(\.invitationRepository) private var invitationRepository

    let maxUsesOptions = [10, 50, 100, 500]
    let expiresDaysOptions = [1, 3, 7, 14, 30]

    func loadCircles() async {
        isLoading = true
        defer { isLoading = false }

        do {
            circles = try await circleRepository.getCircles()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func createInvitation() async -> InvitationDetail? {
        isCreating = true
        defer { isCreating = false }

        do {
            let invitation = try await invitationRepository.createInvitation(
                circleId: selectedCircleId,
                maxUses: maxUses,
                expiresDays: expiresDays
            )
            createdInvitation = invitation
            return invitation
        } catch {
            errorMessage = error.localizedDescription
            return nil
        }
    }
}

File: Android/feature/feature-invitation/src/main/java/com/youkong/feature/invitation/viewmodel/CreateInvitationViewModel.kt
package com.youkong.feature.invitation.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.youkong.core.domain.model.CircleDetail
import com.youkong.core.domain.model.Invitation
import com.youkong.core.domain.repository.CircleRepository
import com.youkong.core.domain.repository.InvitationRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class CreateInvitationUiState(
    val circle: CircleDetail? = null,
    val expireDays: Int = 7,
    val maxUses: Int = 100,
    val isLoading: Boolean = false,
    val isCreating: Boolean = false,
    val createdInvitation: Invitation? = null,
    val error: String? = null,
)

@HiltViewModel
class CreateInvitationViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val circleRepository: CircleRepository,
    private val invitationRepository: InvitationRepository,
) : ViewModel() {

    private val circleId: String? = savedStateHandle["circleId"]

    private val _uiState = MutableStateFlow(CreateInvitationUiState())
    val uiState: StateFlow<CreateInvitationUiState> = _uiState.asStateFlow()

    init {
        if (circleId != null) {
            loadCircle()
        }
    }

    private fun loadCircle() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            circleRepository.getCircle(circleId!!)
                .onSuccess { circle ->
                    _uiState.update { it.copy(isLoading = false, circle = circle) }
                }
                .onFailure { e ->
                    _uiState.update {
                        it.copy(isLoading = false, error = e.message ?: "加载失败")
                    }
                }
        }
    }

    fun setExpireDays(days: Int) {
        _uiState.update { it.copy(expireDays = days) }
    }

    fun setMaxUses(uses: Int) {
        _uiState.update { it.copy(maxUses = uses) }
    }

    fun createInvitation() {
        viewModelScope.launch {
            _uiState.update { it.copy(isCreating = true, error = null) }

            invitationRepository.createInvitation(
                circleId = circleId,
                maxUses = _uiState.value.maxUses,
                expiresDays = _uiState.value.expireDays,
            )
                .onSuccess { invitation ->
                    _uiState.update {
                        it.copy(isCreating = false, createdInvitation = invitation)
                    }
                }
                .onFailure { e ->
                    _uiState.update {
                        it.copy(isCreating = false, error = e.message ?: "创建失败")
                    }
                }
        }
    }
}
